zenoh/src/net/routing/dispatcher/pubsub.rs

```diff
@@ -0,0 +1,621 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use super::face::FaceState;
+use super::resource::{DataRoutes, Direction, PullCaches, Resource};
+use super::tables::{NodeId, Route, RoutingExpr, Tables, TablesLock};
+use crate::net::routing::hat::HatTrait;
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::sync::{Arc, MutexGuard};
+use zenoh_core::zread;
+use zenoh_protocol::core::key_expr::{keyexpr, OwnedKeyExpr};
+use zenoh_protocol::network::declare::subscriber::ext::SubscriberInfo;
+use zenoh_protocol::network::declare::Mode;
+use zenoh_protocol::{
+    core::{WhatAmI, WireExpr},
+    network::{declare::ext, Push},
+    zenoh::PushBody,
+};
+use zenoh_sync::get_mut_unchecked;
+
+pub(crate) fn declare_subscription(
+    hat_code: &MutexGuard<'_, Box<dyn HatTrait + Send + Sync>>,
```

Comment:
It seems it is not necessary to pass a reference to the `MutexGuard`, a `&dyn HatTrait + Send + Sync` could be passed instead. Is it more a convention rather than a real need?